### PR TITLE
[SPARK-36461][SQL] Enable ObjectHashAggregate for more Aggregate functions

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

Enabing more `ObjectHashAggregate` for more aggregate functions, as it has better performance compared to `SortAggregate` according to current benchmark.

### Why are the changes needed?

We are pretty conservative with `ObjectHashAggregate` right now. Except for enabling it by configuring, there is also a fallback threshold. With the fallback threshold, it looks like that even `ObjectHashAggregate` is bad at memory estimation, it is unlikely to OOM easily. For now the users cannot work with `ObjectHashAggregate` for these aggregation functions even they enable the config. That's said, `ObjectHashAggregate` is not much able to be applicable under these limitations. As we are enough conservative with fallback threshold value, it is also disable-able using the config, the limitations on applicable aggregation functions look not necessary.

### Does this PR introduce _any_ user-facing change?

Yes. Some aggregations will be applied with `ObjectHashAggregate` after this change.

### How was this patch tested?

Added unit tests.

### CODE REVIEW

**Comment by viirya** [comment]
Looked at `ObjectHashAggregate `, wondering why we don't enable some Aggregate functions on it. I ran it though existing tests, seems it is okay. Wanted to check with CI.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ObjectHashAggregateSuite.scala [hunk 1]
```diff
@@ -223,6 +257,7 @@ class ObjectHashAggregateSuite
 
     val allAggs = Seq(
       "typed" -> typed,
+      "imper_agg" -> imper_agg,
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ObjectHashAggregateSuite.scala:260]: nit but maybe just `imperAgg` to follow other codes here.

**Comment by tanelk** [comment]
A short discussion on this topic from a while back: https://github.com/apache/spark/pull/32057#issuecomment-818743773

**Comment by viirya** [comment]
Thanks for referring to previous discussion. I agree with @cloud-fan. However, currently it is not optional for users. I'm thinking if we can offer users options here?

**Comment by viirya** [comment]
Actually we already have `spark.sql.execution.useObjectHashAggregateExec` to control the use of `ObjectHashAggregateExec`. It sounds a bit redundant to add another config here.

I also think the current condition `TypedImperativeAggregate` is already more strict than others like `ImperativeAggregate` and `DeclarativeAggregate`.

**Comment by viirya** [comment]
cc @cloud-fan What do you think? I think we are conservative enough with the fallback threshold. With the fallback threshold, it looks like that even `ObjectHashAggregateExec` is bad at memory estimation, it is unlikely to OOM easily. For now the users cannot work with `ObjectHashAggregateExec` for these cases even they enable the config.

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Pull request closed without merging**